spec/expression.dd

```diff
@@ -2417,21 +2412,51 @@ $(GNAME AssertArguments):
     of the program is undefined.)
 
     $(IMPLEMENTATION_DEFINED Whether the first $(I AssertExpression) is evaluated
-    or not at runtime is typically set with a compiler switch. If it is not evaluated,
+    or not (at runtime) is typically set with a compiler switch. If it is not evaluated,
     any side effects specified by the $(I AssertExpression) may not occur.
-    The behavior if the first $(I AssertExpression) is evaluated and is false
-    is also typically set with a compiler switch and may include these options:
+    The behavior when the first $(I AssertExpression) evaluates to `false`
+    is also typically set with a compiler switch, and may include these options:
     $(OL
-        $(LI continuing execution)
-        $(LI immediately halting via execution of a special CPU instruction)
-        $(LI aborting the program)
-        $(LI calling the assert failure function in the corresponding C
+        $(LI Immediately halting via execution of a special CPU instruction)
+        $(LI Aborting the program)
+        $(LI Calling the assert failure function in the corresponding C
         runtime library)
-        $(LI throwing the $(D AssertError) exception in the D runtime library)
+        $(LI Throwing the $(D AssertError) exception in the D runtime library)
+    )
+    )
+
+    $(BEST_PRACTICE
+    $(OL
+        $(LI Do not have side effects in either $(I AssignExpression) that subsequent code
+        depends on.)
+        $(LI $(I AssertExpression)s are intended to detect bugs in the program.
+        Do not use them for detecting input or environmental errors.)
+        $(LI Do not attempt to resume normal execution after an $(I Assert Failure).)
+    )
+    )
+
+$(H4 $(LNAME2 assert-ct, Compile-time Evaluation))
+
+    $(P If the first $(I AssignExpression) consists entirely of compile time constants,
+    and evaluates to `false`, it is a special case - it
+    signifies that subsequent statements are unreachable code.
+    Compile Time Function Execution (CTFE) is not attempted.
+    )
+
+    $(P The implementation may handle the case of the first $(I AssignExpression) evaluating to `false`
+    at compile time differently - even when other `assert`s are ignored,
+    it may still generate a $(D HLT) instruction or equivalent.
     )
-    If the optional second $(I AssignExpression) is provided, the implementation may
-    evaluate it and print the resulting message upon assert failure:
 
+    $(P See also: $(DDSUBLINK spec/version, static-assert, `static assert`).)
+
+$(H4 $(LNAME2 assert-message, Assert Message))
+
+    $(P The second $(I AssignExpression), if present, must be implicitly
+        convertible to type $(D const(char)[]).
```

Comment:
assert can now have multiple parameters, each of which is converted to string and concatenated. So this sentence should start with "any following AssignExpression, if present, ...

Reply:
@Lyratelle No that only works with `static assert`. Otherwise `dmd` would have to link in a runtime function to e.g. convert int to string.